### PR TITLE
Make rrule fast forwarding stable

### PR DESCRIPTION
**Pull request by fosterseth**
##### SUMMARY

By stable, we mean future occurrences of the rrule should be the same before and after the fast forward operation.

The problem before was that we were fast forwarding to *exactly* 7 days ago. For some rrules, this does not retain the old occurrences (for example, freq=HOURLY and INTERVAL=23). Thus, jobs would launch at unexpected times any time the schedule would fast forward.

This change makes sure we fast forward in increments of the rrule INTERVAL (converted to seconds), thus the new dtstart should be in the occurrence list of the old rrule.

#### DETAIL

It is important that each rrule in the rruleset gets fast forwarded independently. For example consider this rruleset

```
DTSTART;TZID=America/New_York:20201118T200000
RRULE:FREQ=HOURLY;INTERVAL=5
RRULE:FREQ=HOURLY;INTERVAL=7
```
Each of those rrules will need its own dtstart, since the occurrences from each rrule do not perfectly overlap.

the API also support exclusion rules, which are basically just rrules. So we can use the same method to fast forward those too.

#### PERFORMANCE

We want to retain the performance gains from introducing fast forwarding to begin with

create a bunch of old schedules

```
for i in range(2000):
    Schedule.objects.create(name=f's{i}', rrule='DTSTART;TZID=America/New_York:20190517T000000 RRULE:FREQ=HOURLY;INTERVAL=7', unified_job_template_id=7)
```

runit will update_computed_fields for each schedule
```
def runit():
    for sch in Schedule.objects.all():
        sch.update_computed_fields()
```

benchmark it
```
cProfile.run("runit()", "bench.txt")
```

before:
image

after:
image

##### ISSUE TYPE

 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME

 - API

### CODE REVIEW

File: awx/main/models/schedules.py [hunk 1]
```diff
@@ -207,23 +245,33 @@ def rrulestr(cls, rrule, fast_forward=True, **kwargs):
...
         # Fast forward is a way for us to limit the number of events in the rruleset
-        # If we are fastforwading and we don't have a count limited rule that is minutely or hourley
+        # If we are fast forwarding and we don't have a count limited rule that is minutely or hourly
         # We will modify the start date of the rule to last week to prevent a large number of entries
         if fast_forward:
             try:
                 # All rules in a ruleset will have the same dtstart value
-                #   so lets compare the first event to now to see if its > 7 days old
+                # so lets compare the first event to now to see if its > 30 days old
+                # we choose > 30 hours because if rrule has FREQ=HOURLY and INTERVAL=23, it will take
```
Comment (AlanCoding) [inline awx/main/models/schedules.py]: Comment says "> 30 days old" and then "> 30 hours". This doesn't look intentional.

Comment (fosterseth) [reply awx/main/models/schedules.py]: oh thanks, fixed

**Comment by PabloHiro** [comment]
> Fast forward won't work for really large intervals. For example if you specify HOURLY and INTERVAL=1200 (that is 50 days worth of time), then we can't fast forward because one CHUNK of the interval doesn't fit in the period of time we are trying to fast forward (window is 30 days). In this case, we'll revert back to the old style of just updating dtstart to 7 days ago. We log a warning, so hopefully the user will update their rrule to be frequency DAYS instead of HOURS.

Can this be an HTTP 400 when a Schedule is created with an "invalid" rrule? Basically preventing this type of rrules to be created in the first place.

Also, I would specify the value of the `fast forward amount` in the logs, so the user does not need to guess the value by trial and error for existing "invalid" rrules

**Comment by pb82** [comment]
Ok, I think I understood what's going on:

1. To avoid computing a large number of events for an rrule with a dtstart way back in the past, we only compute from one week ago.
2. To not mess up the interval, we can't just subtract 7 days, we have to go back an amount of time that is dividable by the interval (that's what `(fast_forward_seconds // interval) * interval` does).

Did I get that right?

File: awx/main/models/schedules.py [hunk 2]
```diff
@@ -207,23 +245,33 @@ def rrulestr(cls, rrule, fast_forward=True, **kwargs):
...
-                            new_start = (now() - datetime.timedelta(days=7)).strftime('%Y%m%d')
-                            # Now we want to repalce the DTSTART:<value>T with the new date (which includes the T)
-                            new_rrule = re.sub('(DTSTART[^:]*):[^T]+T', r'\1:{0}T'.format(new_start), rrule)
+                            try:
+                                new_start = fast_forward_date(rule)
+                            except RuntimeError as e:
+                                logger.warning(e)
+                                # fallback to setting dtstart to 7 days ago, but this has the consequence of
+                                # occurrences not matching the old occurrences.
+                                new_start = now() - datetime.timedelta(days=7)
```
Comment (pb82) [inline awx/main/models/schedules.py]: as @PabloHiro mentioned, this would lead to the exact problem that is being fixed. Can we just reject the rrule here?

Comment (PabloHiro) [reply awx/main/models/schedules.py]: I wonder how we should handle the cases where a user has these "invalid" rules already. We can prevent further invalid rules, but this code has a solution to handle the existing ones.

**Review by djyasin — APPROVED** [review]

**Review by AlanCoding — APPROVED** [review]

**Comment by pb82** [comment]
@fosterseth can this be merged? The failures seem to be about code coverage.

File: awx/main/models/schedules.py [hunk 3]
```diff
@@ -35,6 +35,45 @@
...
+
+def fast_forward_date(rrule):
+    '''
+    Utility to fast forward an rrule, maintaining consistency in the resulting
+    occurrences
+    Fast forwards the rrule to 7 days ago
+    Returns a datetime object
+    '''
+    if not rrule._freq in (dateutil.rrule.HOURLY, dateutil.rrule.MINUTELY):
+        raise RuntimeError("Cannot fast forward rrule, frequency must be HOURLY or MINUTELY")
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: This seems to be semantically closer to a `ValueError` than to a `RuntimeError` as it is function input validation, not a runtime check. Perhaps, it's worth changing?
```suggestion
        raise ValueError(f"Cannot fast forward rrule, frequency must be HOURLY or MINUTELY, but got {rrule._freq !r}")
```

File: awx/main/models/schedules.py [hunk 4]
```diff
@@ -35,6 +35,45 @@
...
+
+
+def fast_forward_date(rrule):
+    '''
+    Utility to fast forward an rrule, maintaining consistency in the resulting
+    occurrences
+    Fast forwards the rrule to 7 days ago
+    Returns a datetime object
+    '''
+    if not rrule._freq in (dateutil.rrule.HOURLY, dateutil.rrule.MINUTELY):
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: `not x in y` is a common mistake. Linters would usually require it to be an `x not in y` instead, as it's clearer what the semantics of the check is. I wouldn't even be sure whether it's `(not x) in y` or `not (x in y)` off the top of my head. Let's improve this a bit:
```suggestion
    if rrule._freq not in {dateutil.rrule.HOURLY, dateutil.rrule.MINUTELY}:
```

File: awx/main/models/schedules.py [hunk 5]
```diff
@@ -35,6 +35,45 @@
...
+        raise RuntimeError("Cannot fast forward rrule, frequency must be HOURLY or MINUTELY")
+
+    interval = rrule._interval if rrule._interval else 1
+    if rrule._freq == dateutil.rrule.HOURLY:
+        interval *= 60 * 60
+    elif rrule._freq == dateutil.rrule.MINUTELY:
+        interval *= 60
+
+    if type(interval) == float and not interval.is_integer():
+        raise RuntimeError("Cannot fast forward rule, interval is a fraction of a second")
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: Same concern here: should this be a `ValueError`?

File: awx/main/models/schedules.py [hunk 6]
```diff
@@ -35,6 +35,45 @@
...
+    if not rrule._freq in (dateutil.rrule.HOURLY, dateutil.rrule.MINUTELY):
+        raise RuntimeError("Cannot fast forward rrule, frequency must be HOURLY or MINUTELY")
+
+    interval = rrule._interval if rrule._interval else 1
+    if rrule._freq == dateutil.rrule.HOURLY:
+        interval *= 60 * 60
+    elif rrule._freq == dateutil.rrule.MINUTELY:
+        interval *= 60
+
+    if type(interval) == float and not interval.is_integer():
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: Is it intentional that `isinstance(interval, float)` is not used here? Do we want to disallow subclasses?
Also, wouldn't it be more correct to do an identity check, as in `type(interval) is float`?

Looking at what's checked, I think that it should be possible to use `rrule._interval` instead and move it to the beginning of the function, having a chance to bail before the computations. This is probably not a huge performance benefit, but it would improve the structure by grouping all the “guard expression”-style checks together.

Additionally, once these checks are close, they could be moved into a helper called smth like `_assert_rrule_is_fast-forwardable()`, which might improve the structure even further.

File: awx/main/models/schedules.py [hunk 7]
```diff
@@ -35,6 +35,45 @@
...
+    # make sure at least one chunk of interval can fit inside the time period we are trying to fast forward
+    if interval > fast_forward_seconds:
+        raise RuntimeError(f"Cannot fast forward rrule {rrule}, interval is greater than the fast forward amount of {fast_forward_seconds} seconds")
+
+    # it is important to fast forward by a number that is divisible by
+    # interval. For example, if interval is 7 hours, we fast forward by 7, 14, 21, etc. hours.
+    # Otherwise, the occurrences after the fast forward might not match the ones before.
+    # x // y is integer division, lopping off any remainder, so that we get the outcome we want.
+    new_start = rrule._dtstart + datetime.timedelta(seconds=(fast_forward_seconds // interval) * interval)
+    return new_start
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: Seeing this code comment, makes me think that this line is hard to follow, and it might not belong on this abstraction layer. I believe that such things can be labeled with descriptive variable names (or function names for that matter), improving the perception.
Wouldn't it be easier to follow if this were to read
```python
return _compute_interval_aligned_datetime(rrule._dtstart, fast_forward_seconds, interval)
```
or
```python
return rrule._dtstart + _compute_interval_aligned_time_shift(fast_forward_seconds, interval)
```
?

File: awx/main/models/schedules.py [hunk 8]
```diff
@@ -207,23 +246,33 @@ def rrulestr(cls, rrule, fast_forward=True, **kwargs):
...
+        # If we are fast forwarding and we don't have a count limited rule that is minutely or hourly
         # We will modify the start date of the rule to last week to prevent a large number of entries
         if fast_forward:
             try:
                 # All rules in a ruleset will have the same dtstart value
-                #   so lets compare the first event to now to see if its > 7 days old
+                # so lets compare the first event to now to see if its > 30 days old
+                # we choose > 30 days because if rrule has FREQ=HOURLY and INTERVAL=23, it will take
+                # at least 23 days before we can fast forward reliably and retain stable occurrences.
+                # since we are fast forwarding to 7 days ago, we check fo rrrules older than (23+7) days
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: `s/rrr/rr/`

File: awx/main/models/schedules.py [hunk 9]
```diff
@@ -207,23 +246,33 @@ def rrulestr(cls, rrule, fast_forward=True, **kwargs):
...
                 first_event = x[0]
-                if (now() - first_event).days > 7:
+                if (now() - first_event).days > 30:
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: May I suggest moving the arbitrary constant of `30` into a variable? Maybe, a class/module-level constant. This piece of data needs a name.

File: awx/main/models/schedules.py [hunk 10]
```diff
@@ -207,23 +246,33 @@ def rrulestr(cls, rrule, fast_forward=True, **kwargs):
...
                         if rule._freq in [dateutil.rrule.MINUTELY, dateutil.rrule.HOURLY] and not rule._count:
                             # hourly/minutely rrules with far-past DTSTART values
                             # are *really* slow to precompute
                             # start *from* one week ago to speed things up drastically
-                            new_start = (now() - datetime.timedelta(days=7)).strftime('%Y%m%d')
-                            # Now we want to repalce the DTSTART:<value>T with the new date (which includes the T)
-                            new_rrule = re.sub('(DTSTART[^:]*):[^T]+T', r'\1:{0}T'.format(new_start), rrule)
+                            try:
+                                new_start = fast_forward_date(rule)
+                            except RuntimeError as e:
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: Since there's nothing wrong with the runtime, this would also need to align with the above suggestions to represent the case with a different exception:
```suggestion
                            except ValueError as val_err:
```

File: awx/main/models/schedules.py [hunk 11]
```diff
@@ -207,23 +246,33 @@ def rrulestr(cls, rrule, fast_forward=True, **kwargs):
...
+                                logger.warning(e)
+                                # fallback to setting dtstart to 7 days ago, but this has the consequence of
+                                # occurrences not matching the old occurrences.
+                                new_start = now() - datetime.timedelta(days=7)
+                            new_start_fmt = new_start.strftime('%Y%m%d')
+                            # Now we want to replace the DTSTART:<value>T with the new date (which includes the T)
+                            new_rrule = re.sub('(DTSTART[^:]*):[^T]+T', r'\1:{0}T'.format(new_start_fmt), rrule)
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: try this
```suggestion
                            new_rrule = re.sub('(?P<dstart>DTSTART[^:]*):[^T]+T', fr'\g<dstart>:{new_start_fmt !s}T', rrule)
```

File: awx/main/tests/unit/utils/test_schedule_fast_forward.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+    # get occurrences of the rrule
+    gen = rule.xafter(new_datetime - datetime.timedelta(days=1), count=100)
+    found_matching_date = False
+    for occurrence in gen:
+        if occurrence == new_datetime:
+            found_matching_date = True
+            break
+
+    assert found_matching_date
```
Comment (webknjaz) [inline awx/main/tests/unit/utils/test_schedule_fast_forward.py]: Any reason it's not
```suggestion
    assert new_datetime in gen
```
?

File: awx/main/tests/unit/utils/test_schedule_fast_forward.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+@pytest.mark.parametrize(
+    'freq, interval',
+    [
+        (MINUTELY, 15),
+        (MINUTELY, 120),
+        (MINUTELY, 60 * 24 * 3),
+        (HOURLY, 7),
+        (HOURLY, 24 * 3),
+    ],
```
Comment (webknjaz) [inline awx/main/tests/unit/utils/test_schedule_fast_forward.py]: (cosmetic changes)
```suggestion
    ('freq', 'interval'),
    (
        pytest.param(MINUTELY, 15, id='every-15-minutes-minutely'),
        pytest.param(MINUTELY, 120, id='every-2-hours-minutely'),
        pytest.param(MINUTELY, 60 * 24 * 3, id='every-3-days-minutely'),
        pytest.param(HOURLY, 7, id='every-7-hours-hourly'),
        pytest.param(HOURLY, 24 * 3, id='every-3-days-hourly'),
    ),
```

File: awx/main/tests/unit/utils/test_schedule_fast_forward.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+)
+def test_error_fast_forward_date(freq, interval, error):
+    dtstart = now() - datetime.timedelta(days=30)
+    rule = rrule(freq=freq, interval=interval, dtstart=dtstart)
+    if error:
+        with pytest.raises(Exception) as e_info:
+            fast_forward_date(rule)
+
+        assert error in e_info.value.args[0]
```
Comment (webknjaz) [inline awx/main/tests/unit/utils/test_schedule_fast_forward.py]: It's best to avoid convoluted logic in tests. Error is always truthy, so the conditional is not really needed. Additionally, the right way of checking the exceptions raised is to narrow the expectation. `pytest.raises()` uses (and `pytest.warns()` or `pytest.deprecated()` for that matter) should always have a `match` argument to be as accurate as possible. If there's different expected exceptions, it's possible to bake them into parametrize. But let's start with the following:
```suggestion
@pytest.mark.parametrize(
    ('freq', 'interval', 'error'),
    (
        pytest.param(MINUTELY, 15.5555, r"^interval is a fraction of a second$", id='fraction-of-sec'),
        pytest.param(MONTHLY, 1, r"^frequency must be HOURLY or MINUTELY$", id='monthly'),
        pytest.param(HOURLY, 24 * 30, r"^interval is greater than the fast forward amount$", id='over-a-month'),
    ),
)
def test_error_fast_forward_date(freq, interval, error):
    dtstart = now() - datetime.timedelta(days=30)
    rule = rrule(freq=freq, interval=interval, dtstart=dtstart)

    with pytest.raises(ValueError, match=error):
        fast_forward_date(rule)
```

I'm also concerned by the use of `now()` as it may contribute to the flakiness of the tests. It's typical to freeze time in such cases.

**Review by webknjaz — COMMENTED** [review]
got a few observations below

File: awx/main/models/schedules.py (hunk 11 above)
Comment (fosterseth) [reply awx/main/models/schedules.py]: no longer needed, using rrule.replace() instead of injecting the dtstart string

File: awx/main/tests/unit/utils/test_schedule_fast_forward.py (hunk 12 above)
Comment (fosterseth) [reply awx/main/tests/unit/utils/test_schedule_fast_forward.py]: good call! I rewrote this block so no longer relevant

File: awx/main/models/schedules.py [hunk 15]
```diff
@@ -35,6 +35,61 @@
 UTC_TIMEZONES = {x: tzutc() for x in dateutil.parser.parserinfo().UTCZONE}
 
 
+def _check_valid_tzid(rrules):
+    for r in rrules:
+        if r._dtstart and r._dtstart.tzinfo is None:
+            raise ValueError('A valid TZID must be provided (e.g., America/New_York)')
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: Since this is not doing `if any(): raise` but has an explicit top-level loop, I think it's important to mention what actually is wrong and where.

Could you add the specific rrule details into the message? And maybe spell out the abbreviations? As in, change “TZID” to “time zone identifier”, for example.

But also, perhaps it'd be better to go through all the rules and then add details on the batch of broken rules?

I'm thinking something along the lines of
```suggestion
def _assert_timezone_id_is_valid(rrules) -> None:
    broken_rrules = [
        rrule for rrule in rrules
        if r._dtstart and r._dtstart.tzinfo is None
    ]
    if not broken_rrules:
        return
    raise ValueError(
        'The following rrules do not a timezone set: '
        f'{broken_rrules}. Please, set it for each '
        '(e.g., America/New_York).',
    ) from None
```

File: awx/main/models/schedules.py (hunk 6 above)
Comment (fosterseth) [reply awx/main/models/schedules.py]: made some recent code changes around this

File: awx/main/models/schedules.py (hunk 10 above)
Comment (fosterseth) [reply awx/main/models/schedules.py]: no longer throwing error

File: awx/main/tests/unit/utils/test_schedule_fast_forward.py (hunk 13 above)
Comment (fosterseth) [reply awx/main/tests/unit/utils/test_schedule_fast_forward.py]: incorporated this pattern in to my tests, although this specific test went away

File: awx/main/tests/unit/utils/test_schedule_fast_forward.py (hunk 14 above)
Comment (fosterseth) [reply awx/main/tests/unit/utils/test_schedule_fast_forward.py]: I replaced usage of now() with a static reference datetime (`REF_DT`)

File: awx/main/models/schedules.py [hunk 16]
```diff
@@ -35,6 +35,64 @@
 UTC_TIMEZONES = {x: tzutc() for x in dateutil.parser.parserinfo().UTCZONE}
 
 
+def _assert_timezone_id_is_valid(rrules) -> None:
+    broken_rrules = [str(rrule) for rrule in rrules if rrule._dtstart and rrule._dtstart.tzinfo is None]
+    if not broken_rrules:
+        return
+    raise ValueError(
+        'The following rrules do not have a valid timezone: ' f'{broken_rrules}. Set a valid timezone ' '(e.g., America/New_York).',
```
Comment (webknjaz) [inline awx/main/models/schedules.py]: Looks like the formatter did the dumb thing and inlined all the implicit concat.
```suggestion
        f'The following rrules do not have a valid timezone: {broken_rrules}. Set a valid timezone (e.g., America/New_York).',
```

File: awx/main/tests/unit/utils/test_schedule_fast_forward.py [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+
+
+@pytest.mark.parametrize(
+    'rrulestr',
+    [
+        'DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=MINUTELY;INTERVAL=5',
+        'DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=HOURLY;INTERVAL=5',
+        'DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=YEARLY;INTERVAL=5',
+        'DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=MINUTELY;INTERVAL=5;WKST=SU;BYMONTH=2,3;BYMONTHDAY=18;BYHOUR=5;BYMINUTE=35;BYSECOND=0',
+        'DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=HOURLY;INTERVAL=5;WKST=SU;BYMONTH=2,3;BYHOUR=5',
```
Comment (webknjaz) [inline awx/main/tests/unit/utils/test_schedule_fast_forward.py]: I think some of these broke my eyes... Hopefully I interpreted them correctly:
```suggestion
        pytest.param('DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=MINUTELY;INTERVAL=5', id='every-5-min'),
        pytest.param('DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=HOURLY;INTERVAL=5', id='every-5-hours'),
        pytest.param('DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=YEARLY;INTERVAL=5', id='every-5-years'),
        pytest.param('DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=MINUTELY;INTERVAL=5;WKST=SU;BYMONTH=2,3;BYMONTHDAY=18;BYHOUR=5;BYMINUTE=35;BYSECOND=0', id='every-5-minutes-at-5:35:00-am-on-the-18th-day-of-feb-or-march-with-week-starting-on-sundays'),
        pytest.param('DTSTART;TZID=America/New_York:20201118T200000 RRULE:FREQ=HOURLY;INTERVAL=5;WKST=SU;BYMONTH=2,3;BYHOUR=5', id='every-5-hours-at-5-am-in-feb-or-march-with-week-starting-on-sundays'),
```

This should improve the pytest test report output and give the strings so much needed labels in code.
